src/processor.cpp

```diff
@@ -0,0 +1,401 @@
+        }
+    }
+}
+
+void processor_data::load( JsonObject &jo, const std::string &src )
+{
+    processors_data.load( jo, src );
+}
+
+void processor_data::check()
+{
+    processors_data.check();
+}
+
+void processor_data::reset()
+{
+    processors_data.reset();
+}
+
+//process namespace
+
+namespace
+{
+
+struct process_data;
+using process_id = string_id<process_data>;
+
+struct process_data {
+    process_data():
+        was_loaded( false ) {};
+
+    process_id id;
+    std::string name;
+    auto components;
+    auto output;
+    int fuel_intake;
+    int duration;
+    //TODO: Skill used
+
+    void load( JsonObject &jo, const std::string &src );
+    void check() const;
+
+};
+
+process_id get_process_id( std::string id )
+{
+    return process_id( id );
+}
+
+generic_factory<process_data> processes_data( "process type", "id" );
+
+}
+
+void process_data::load( JsonObject &jo, const std::string & )
+{
+    mandatory( jo, was_loaded, "name", name );
+    //TODO: modify to allow lists of components
+    mandatory( jo, was_loaded, "components", components,  string_id_reader<itype_id> {});
+    mandatory( jo, was_loaded, "output", output );
+    mandatory( jo, was_loaded, "fuel_intake", fuel_intake );
+    mandatory( jo, was_loaded, "duration", duration );
+}
+
+void process_data::check() const
+{
+    const process_id proc_id( id.str() );
+
+    if( !proc_id.is_valid() ) {
+        debugmsg( "There is no process with the name %s", id.c_str() );
+    if( !components.is_valid() ) {
+        debugmsg( "Invalid components \"%s\" in \"%s\".", components.c_str(), id.c_str() );
+    }
+    if( !output.is_valid() ) {
+        debugmsg( "Invalid output \"%s\" in \"%s\".", output.c_str(), id.c_str() );
+    }
+    if( !fuel_intake.is_valid() ) {
+        debugmsg( "Invalid fuel intake \"%s\" in \"%s\".", fuel_intake.c_str(), id.c_str() );
+    }
+    if( !duration.is_valid() ) {
+        debugmsg( "Invalid duration \"%s\" in \"%s\".", duration.c_str(), id.c_str() );
+    }
+}
+
+void process_data::load( JsonObject &jo, const std::string &src )
+{
+    processes_data.load( jo, src );
+}
+
+void process_data::check()
+{
+    processes_data.check();
+}
+
+void process_data::reset()
+{
+    processes_data.reset();
+}
+
+void interact_with_processor( const tripoint &examp, player &p )
+{
+    //Get processor id at examp
+    processor_id pid = get_processor_id( examp );
+    //Do I need the debugmsg-es here or do they get checked in process_data::load?
+    if( !processors_data.is_valid( pid ) ) {
+        debugmsg( "Examined %s has action interact_with_processor, but has no processor data associated", g->m.furn( examp ).id().c_str() );
+        return;
+    }
+    //Get data associated with said processor id
+    const processor_data &processor = processors_data.obj( pid );
+    std::vector<auto> possible_processes;
+    for( auto &p : processor.processes() ) {
+        if( p.is_valid() ){
```

Comment:
Here however you shouldn't have validity check. That should be handled earlier - a processor should bug out at verification if it gets an invalid process.